File: src/main/java/wraith/fabricaeexnihilo/json/recipe/SieveRecipeJson.java
package wraith.fabricaeexnihilo.json.recipe;

import com.google.gson.*;
import wraith.fabricaeexnihilo.recipe.util.FluidIngredient;
import wraith.fabricaeexnihilo.recipe.util.ItemIngredient;
import wraith.fabricaeexnihilo.api.recipes.SieveRecipe;
import wraith.fabricaeexnihilo.json.BaseJson;

import java.lang.reflect.Type;

public final class SieveRecipeJson extends BaseJson<SieveRecipe> {

    private SieveRecipeJson() {}

    public static final SieveRecipeJson INSTANCE = new SieveRecipeJson();

    @Override
    public SieveRecipe deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        var obj = json.getAsJsonObject();
        return new SieveRecipe(
                ItemIngredient.fromJson(obj.get("mesh")),
                obj.get("fluid") != null ? FluidIngredient.fromJson(obj.get("fluid")) : FluidIngredient.EMPTY,
                ItemIngredient.fromJson(obj.get("sievable")),
                context.deserialize(obj.get("loot"), LOOTABLE_LIST_TYPE_TOKEN)
        );
    }

    @Override
    public JsonElement serialize(SieveRecipe src, Type typeOfSrc, JsonSerializationContext context) {
        var obj = new JsonObject();
        obj.add("mesh", src.mesh().toJson());
        if(!src.fluid().isEmpty()) {
            obj.add("fluid", src.fluid().toJson());
        }
        obj.add("sievable", src.sievable().toJson());
        obj.add("loot", context.serialize(src.loot()));
        return obj;
    }

}

File: src/main/kotlin/exnihilofabrico/json/SieveRecipeJson.kt
package exnihilofabrico.json

import com.google.gson.*
import com.google.gson.reflect.TypeToken
import exnihilofabrico.api.crafting.FluidIngredient
import exnihilofabrico.api.crafting.ItemIngredient
import exnihilofabrico.api.crafting.Lootable
import exnihilofabrico.api.recipes.SieveRecipe
import java.lang.reflect.Type

object SieveRecipeJson: JsonDeserializer<SieveRecipe>, JsonSerializer<SieveRecipe> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): SieveRecipe {
        val obj = json.asJsonObject
        return SieveRecipe(
            ItemIngredient.fromJson(obj["mesh"], context),
            if(obj.get("fluid") != null)
                FluidIngredient.fromJson(obj["fluid"], context)
            else
                FluidIngredient.EMPTY
            ,
            ItemIngredient.fromJson(obj["sievable"], context),
            context.deserialize(obj["loot"], object: TypeToken<MutableCollection<Lootable>>(){}.type)
        )
    }

    override fun serialize(src: SieveRecipe, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()
        obj.add("mesh", src.mesh.toJson(context))
        if(!src.fluid.isEmpty())
            obj.add("fluid", src.fluid.toJson(context))
        obj.add("sievable", src.sievable.toJson(context))
        obj.add("loot", context.serialize(src.loot))
        return obj
    }

}